File: java_solutions/3257.maximum-value-sum-by-placing-three-rooks-ii.java
/*
 * @lc app=leetcode id=3257 lang=java
 *
 * [3257] Maximum Value Sum by Placing Three Rooks II
 */

class Solution { public long maximumValueSum(int[][] board) { return ((Function<int[][], Long>) arr -> IntStream.range(0, arr.length).boxed().flatMap(i -> IntStream.range(i + 1, arr.length).filter(j -> arr[i][1] != arr[j][1] && arr[i][2] != arr[j][2]).boxed().flatMap(j -> IntStream.range(j + 1, arr.length).filter(k -> arr[i][1] != arr[k][1] && arr[j][1] != arr[k][1] && arr[i][2] != arr[k][2] && arr[j][2] != arr[k][2]).mapToObj(k -> (long)arr[i][0] + (long)arr[j][0] + (long)arr[k][0]))).reduce(-4000000000L, Math::max)).apply(((Function<Object[], int[][]>) state -> IntStream.range(0, board.length).boxed().flatMap(i -> IntStream.range(0, board[0].length).mapToObj(j -> new int[]{board[i][j], i, j}).sorted((a, b) -> Integer.compare(b[0], a[0])).limit(3)).sorted((a, b) -> Integer.compare(b[0], a[0])).filter(a -> ((int[])state[0])[a[2]] < 3 ? (((int[])state[0])[a[2]]++ >= 0) : false).limit(100).toArray(int[][]::new)).apply(new Object[]{new int[board[0].length]})); } }

File: kotlin_solutions_1/3257.maximum-value-sum-by-placing-three-rooks-ii.kt
/*
 * @lc app=leetcode id=3257 lang=java
 *
 * [3257] Maximum Value Sum by Placing Three Rooks II
 */
class Solution { fun maximumValueSum(board: Array<IntArray>): Long = board.flatMapIndexed { r, row -> row.mapIndexed { c, v -> longArrayOf(v.toLong(), r.toLong(), c.toLong()) }.sortedByDescending { it[0] }.take(3) }.sortedByDescending { it[0] }.take(200).let { c -> c.indices.fold(Long.MIN_VALUE) { m, i -> (i + 1 until c.size).fold(m) { m2, j -> (j + 1 until c.size).fold(m2) { m3, k -> if (c[i][1] != c[j][1] && c[i][1] != c[k][1] && c[j][1] != c[k][1] && c[i][2] != c[j][2] && c[i][2] != c[k][2] && c[j][2] != c[k][2]) maxOf(m3, c[i][0] + c[j][0] + c[k][0]) else m3 } } } } }